### ISSUE TITLE
Search - (edited) label disappears after refreshing Search page

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#50096]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
 
**Version Number:** 9.0.43-1
**Reproducible in staging?:**
**Reproducible in production?:**
**Email or phone of affected tester (no customers):** applausetester+pso@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to DM.
3. Send a message.
4. Go to Search > Chat.
5. Click on the message sent in Step 3.
6. Right click on the message > Edit comment.
7. Edit the comment and save it.
8. Refresh Chat page.

## Expected Result:
(edited) label will persist after refreshing Search page.

## Actual Result:
(edited) label disappears after refreshing Search page.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/473d0643-2756-424d-9a60-383c1205a1ac

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #wave-control

**Comment by huult** [comment]
I believe this is a backend issue because the response from the Search API does not include isEdit in the Onyx data. When this response is saved to the snapshot, it removes isEdit from the snapshot data, causing the issue.

**Comment by CortneyOfstad** [comment]
Still need to test this to confirm it can be recreated! Working on that ASAP

**Comment by CortneyOfstad** [comment]
After fresh testing, I was FINALLY able to get it to happen consistently, so getting eyes on this!

![2024-10-08_09-12-30 (1)](https://github.com/user-attachments/assets/70942de7-6ab5-4d6f-a701-295c5329e773)

**Comment by hungvu193** [comment]
Sounds like a BE issue. I'll review this one in my morning 🥹

**Comment by hungvu193** [comment]
I can reproduce, reviewing shortly

**Comment by hungvu193** [comment]
I think this one can be external. Waiting for proposals!

**Comment by hungvu193** [comment]
I think this is BE issue, `isEdited` didn't return from the BE. Can you please add Internal label? @CortneyOfstad

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Search - (edited) label disappears after refreshing Search page 
 
### What is the root cause of that problem?

The backend does not return `isEdited` in the API response. As a result, the `snapshot` is [updated with the response](https://github.com/Expensify/App/blob/main/src/libs/Middleware/SaveResponseInOnyx.ts#L35) without `isEdited` after the `Search` call is successful, which causes this issue.

- <img width="1508" alt="Screenshot 2024-10-11 at 14 32 15" src="https://github.com/user-attachments/assets/3d7cd565-c2af-483e-b76d-641cd07101e3">


### What changes do you think we should make in order to solve the problem?
I suggest two options to fix this issue:

#### Option 1: We can use the `message` data in `reportAction` from `Onyx` to update the `reportAction` in `snapshot_` because the `reportActions` from `Onyx` includes `isEdited` and is up to date. Something like that:

1. Create `SnapshotConnection.ts` like [ReportConnection](https://github.com/Expensify/App/blob/main/src/libs/ReportConnection.ts) to get all `snapshot_`
```diff
+           let allSnapshot;
+             Onyx.connect({
+                   key: ONYXKEYS.COLLECTION.SNAPSHOT,
+                   waitForCollectionCallback: true,
+                   callback: (value) => {
+                          allSnapshot = value;
+                  },
+            });

+  function getAllSnapshot() {
+    return allSnapshot;
+ }
```
2. Update the `reportAction` data for `ONYXKEYS.COLLECTION.SNAPSHOT`.
```diff
function getOnyxLoadingData(hash: number): {optimisticData: OnyxUpdate[]; finallyData: OnyxUpdate[]; successData: OnyxUpdate[]} {
  // .src/libs/actions/Search.ts#L30
+    const allSnapshot = SnapshotConnection.getAllSnapshot();
+    const allReportActions = ReportActionsConnection.getAllReportActions();
+    const allSnapshotData = allSnapshot ? allSnapshot[`${ONYXKEYS.COLLECTION.SNAPSHOT}${hash}`]?.data : [];

 // .src/libs/actions/Search.ts#L37
+ const successData: OnyxUpdate[] = [];

+    if (allSnapshotData) {
+        let reportActions: Record<string, any> = {};

+        Object.keys(allSnapshotData).forEach((item) => {
+            if (!item.startsWith(ONYXKEYS.COLLECTION.REPORT_ACTIONS) || !allReportActions) {
+                return;
+            }

+            const allSnapshotDataItem = allSnapshotData[item];
+            const reportActionItem = allReportActions[item];

+            if (reportActionItem) {
+                Object.keys(allSnapshotDataItem).forEach((reportActionID) => {
+                    if (reportActionID in reportActionItem) {
+                        allSnapshotDataItem[reportActionID].message = reportActionItem[reportActionID].message;
+                    }
+                });
+            }

+            reportActions[item] = allSnapshotDataItem;
+        });

+        successData.push({
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.SNAPSHOT}${hash}`,
+            value: {
+                data: reportActions,
+            },
+        });
+    }
}  

 // .src/libs/actions/Search.ts#L144
- API.write(WRITE_COMMANDS.SEARCH, {hash: queryJSON.hash, jsonQuery}, {optimisticData, finallyData});
+ if (successData && successData.length) {
+        API.write(WRITE_COMMANDS.SEARCH, {hash: queryJSON.hash, jsonQuery}, {optimisticData, finallyData, successData});
+    } else {
+        API.write(WRITE_COMMANDS.SEARCH, {hash: queryJSON.hash, jsonQuery}, {optimisticData, finallyData});
+    }
```

#### Option 2: The backend must be updated to return `isEdited`, if applicable.

<summary>
<details>POC

https://github.com/user-attachments/assets/a5c4b50e-5a23-425f-b016-194a07f90256

</details>
</summary>

**Comment by huult** [comment]
@hungvu193 , Could you please review my proposal? If we don't update from the backend, we can use the `message` from `ReportAction` in `Onyx`, which is the first option I suggest. If you agree with this option, I will continue testing and create a test branch soon. The second option is to update from the backend side.

**Comment by hungvu193** [comment]
@huult I don't think looping over all snapshot data to update the `reportActions` is a good idea. I still think BE should fix this.

**Comment by huult** [comment]
@hungvu193 , Yes, thank you for your feedback. If the backend change is complex, we can address it from the frontend.

**Comment by CortneyOfstad** [comment]
Hey @mallenexpensify! I am heading OoO this afternoon (10/15 to 10/23) so reassigning this in the meantime to keep it moving! Thanks!

**Comment by mallenexpensify** [comment]
@hungvu193 can you add 🎀 to assign an internal engineer to confirm they agree it's a bug that needs to be fixed on the backend?  (they can unassign after)

**Comment by hungvu193** [comment]
🎀 👀 🎀

**Comment by pecanoro** [comment]
Yes, it seems we send a smaller version of the action compared to the normal one and isEdited is missing. I am not fully sure how search works on the BE but I am going to take a quick lool

**Comment by pecanoro** [comment]
Found the bug

**Comment by CortneyOfstad** [comment]
I'm back from OoO – thanks for holding down the fort @mallenexpensify!

**Comment by trjExpensify** [comment]
Removing the `Hot pick` label because Rocio picked this up. What remains for this issue?

**Comment by trjExpensify** [comment]
Cool, sounds good. I'll leave that to the BZ assigned. 👍

**Comment by pecanoro** [comment]
@CortneyOfstad Can you double check the bug is gone after fixing it in the back-end?

**Comment by CortneyOfstad** [comment]
Yep! Testing now @pecanoro!

**Comment by CortneyOfstad** [comment]
Just tested and it works! The `edited` label stayed even after refreshing a few times! 

@hungvu193 can you confirm if this needs any regression tests? Once that is confirmed, I can do the Payment Summary and get this closed out 👍 

Thanks!

**Comment by hungvu193** [comment]
Heyy @CortneyOfstad, This is BE issue so I don't think there's payment summary or regression tests are needed. Feel free to close 😄
